### PR TITLE
samples: add multiprotocol BLE and 802.15.4 RPMSG sample

### PR DESCRIPTION
**Pull request by e-rk**
This change introduces multiprotocol RPMsg sample for nRF5340 and enables
the build of that sample as a child image when both IEEE 802.15.4 radio driver and
the Bluetooth SoftDevice controller is enabled.

### CODE REVIEW

File: samples/nrf5340/multiprotocol_rpmsg/README.rst [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+You might need to adjust the Kconfig configuration of this sample to make it compatible with the peer application.
+For example, :option:`CONFIG_BT_MAX_CONN` must be equal to the maximum number of connections supported by the peer application.
+
+Refer to :ref:`bluetooth-samples` for general information about Bluetooth samples.
+
+Dependencies
+************
+
+*Missing dependencies here*
```
Comment (FrancescoSer) [inline samples/nrf5340/multiprotocol_rpmsg/README.rst:46]: @e-rk Please add the missing dependencies here

File: samples/nrf5340/multiprotocol_rpmsg/README.rst [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+The sample supports the following development kits:
+
+.. table-from-rows:: /includes/sample_board_rows.txt
+   :header: heading
+   :rows: nrf5340dk_nrf5340_cpunet
+
+Building and Running
+********************
+
+This sample can be found under :zephyr_file:`samples/bluetooth/multiprotocol` in the Zephyr tree.
```
Comment (FrancescoSer) [inline samples/nrf5340/multiprotocol_rpmsg/README.rst:23]: @e-rk  Can you fix the location mentioning the correct NCS location?

**Review by FrancescoSer — CHANGES_REQUESTED** [review]

File: samples/CMakeLists.txt [hunk 3]
```diff
@@ -56,17 +56,27 @@ if (CONFIG_SECURE_BOOT)
 endif()
 
 # Special handling for the nRF53 when using RPMSG HCI.
-# Automatically include the hci_rpmsg sample as child image, and change
+# Automatically include the hci_rpmsg sample or multiprotocol_rpmsg as child
+# image (depending on the number of protocols used), and change
```
Comment (hubertmis) [inline samples/CMakeLists.txt:60]:
```suggestion
# Automatically include the multiprotocol_rpmsg as child image when both RPMSG HCI
# and RPMSG 802.15.4 Serialization are enabled or the hci_rpmsg sample when
# only RPMSG HCI is enabled. For any of these samples change
```
Right now the comment is ambiguous. I would clarify it.

File: samples/nrf5340/multiprotocol_rpmsg/README.rst [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+###################
+
+Overview
+********
+
+This sample exposes :ref:`softdevice_controller` and IEEE 802.15.4 radio driver support to the nRF5340 application CPU using RPMsg transport which is a part of `OpenAMP <https://github.com/OpenAMP/open-amp/>`__.
+
+The sample requires a nRF5340 development kit. The
+firmware must be flashed to the nRF5340 network core.
+This sample is compatible with the HCI RPMsg driver provided by Zephyr's Bluetooth :ref:`bt_hci_drivers` core.
```
Comment (hubertmis) [inline samples/nrf5340/multiprotocol_rpmsg/README.rst:13]: Should we state compatibility with 802.15.4 counterpart as well? Something like:
...and nRF 802.15.4 Serialization Host module

File: samples/nrf5340/multiprotocol_rpmsg/README.rst [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+   This command will flash both application and network core firmwares to the nRF5340 SoC.
+
+#. Run :ref:`sockets-echo-client-sample` in IEEE 802.15.4 variant on a second board.
+
+#. |connect terminal|
+
+   Note that on the nRF5340 DK has multiple UART instances, so the correct port must be identified.
+
+#. Observe that packets are exchanged between the devices and the nRF5340 advertises over Bluetooth.
```
Comment (hubertmis) [inline samples/nrf5340/multiprotocol_rpmsg/README.rst:60]: I'm confused. One of points above requires enabled 802.15.4 peer and here I'm supposed to observe Bluetooth traffic?

Oh, I see now. I think it should be clarified that "IEEE 802.15.4 packets are exchanges" and observing advertisements should be done with an external tool like smartphone.

File: samples/nrf5340/multiprotocol_rpmsg/prj.conf [hunk 6]
```diff
@@ -0,0 +1,21 @@
+CONFIG_LOG=y
+
+CONFIG_RPMSG_SERVICE=y
+CONFIG_RPMSG_SERVICE_MODE_REMOTE=y
```
Comment (hubertmis) [inline samples/nrf5340/multiprotocol_rpmsg/prj.conf:4]: I think it's enabled indirectly by `NRF_802154_SER_RADIO` and BT HCI RPMSG options. Isn't it?

File: samples/nrf5340/multiprotocol_rpmsg/sample.yaml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+sample:
+  description: Allows Zephyr to provide Bluetooth and 802.15.4 connectivity
```
Comment (hubertmis) [inline samples/nrf5340/multiprotocol_rpmsg/sample.yaml:2]:
```suggestion
  description: Allows Zephyr to provide Bluetooth and IEEE 802.15.4 connectivity
```

File: samples/nrf5340/multiprotocol_rpmsg/src/main.c [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+#include <arch/cpu.h>
+#include <sys/byteorder.h>
+#include <logging/log.h>
+#include <sys/util.h>
+#include <drivers/ipm.h>
+
+#include <openamp/open_amp.h>
+#include <metal/sys.h>
+#include <metal/device.h>
+#include <metal/alloc.h>
```
Comment (hubertmis) [inline samples/nrf5340/multiprotocol_rpmsg/src/main.c:22]: I hope these headers do not have to be included in this sample

**Review by hubertmis — COMMENTED** [review]
I like this sample!
I've found some areas to possibly improve

File: samples/nrf5340/multiprotocol_rpmsg/prj.conf (hunk 6 above)
Comment (e-rk) [reply samples/nrf5340/multiprotocol_rpmsg/prj.conf:4]: Looks like that's the case. Removed.

**Review by hubertmis — APPROVED** [review]
:+1:

**Review by FrancescoSer — APPROVED** [review]

**Review by LuDuda — APPROVED** [review]

**Comment by e-rk** [comment]
Branch has been rebased. Are we good with merging https://github.com/nrfconnect/sdk-nrfxlib/pull/366 and https://github.com/nrfconnect/sdk-zephyr/pull/445?

**Pull request merged**